File: 0052-n-queens-ii/0052-n-queens-ii.java
class Solution {
    public int totalNQueens(int n) {
        int[] queens = new int[n]; 
        Arrays.fill(queens, -1); 
        return solveNQueens(n, queens, 0);
    }

    private int solveNQueens(int n, int[] queens, int row) {
        if (row == n) {
            return 1; 
        }

        int count = 0;

        for (int col = 0; col < n; col++) {
            if (isValid(queens, row, col)) {
                queens[row] = col;
                count += solveNQueens(n, queens, row + 1);
                queens[row] = -1; 
            }
        }

        return count;
    }

    private boolean isValid(int[] queens, int row, int col) {
        for (int prevRow = 0; prevRow < row; prevRow++) {
            int prevCol = queens[prevRow];

            if (prevCol == col || prevCol - prevRow == col - row || prevCol + prevRow == col + row) {
                return false;
            }
        }

        return true;
    }
}

File: 0052-n-queens-ii/0052-n-queens-ii.py
class Solution:
    def totalNQueens(self, n: int) -> int:
        def is_valid(queens, row, col):
            for prev_row in range(row):
                prev_col = queens[prev_row]

                if prev_col == col or prev_col - prev_row == col - row or prev_col + prev_row == col + row:
                    return False
            return True

        def solve_nqueens(queens, row):
            if row == n:
                self.count += 1
                return

            for col in range(n):
                if is_valid(queens, row, col):
                    queens[row] = col
                    solve_nqueens(queens, row + 1)

        queens = [-1] * n
        self.count = 0
        solve_nqueens(queens, 0)
        return self.count
